File: plugins/reporting/skills/devexpress-reporting-wpf/examples/quickstart.cs
// DevExpress WPF Reporting — Quickstart (C#)
// Demonstrates: PrintHelper (modal preview), DocumentPreviewControl + MVVM, ReportDesigner

using System;
using System.Windows;
using DevExpress.Xpf.Printing;
using DevExpress.Xpf.Reports.UserDesigner;
using DevExpress.XtraReports.UI;
using DevExpress.XtraPrinting.Caching;
using DevExpress.Mvvm;
using DevExpress.Mvvm.DataAnnotations;

// ------------------------------------------------------------------
// 1. PrintHelper — Ribbon preview (simplest, no XAML changes)
// ------------------------------------------------------------------
public partial class MainWindow : Window {
    private void btnPreview_Click(object sender, RoutedEventArgs e) {
        var report = new SalesReport { DataSource = SampleData.Get() };
        PrintHelper.ShowRibbonPrintPreviewDialog(this, report);
    }

    private void btnDesign_Click(object sender, RoutedEventArgs e) {
        // Requires WPF subscription
        var designerWindow = new DesignerWindow();
        designerWindow.Show();
    }
}

// ------------------------------------------------------------------
// 2. MVVM ViewModel for DocumentPreviewControl
//
// XAML (in MainWindow.xaml):
//   <Window DataContext="{dxmvvm:ViewModelSource Type=local:ReportViewModel}">
//       <dxp:DocumentPreviewControl RequestDocumentCreation="True"
//                                   DocumentSource="{Binding Report}" />
//   </Window>
// ------------------------------------------------------------------
[POCOViewModel]
public class ReportViewModel : ViewModelBase {
    public virtual XtraReport Report { get; protected set; }

    protected override void OnInitialized() {
        LoadReport();
    }

    public void LoadReport() {
        Report = new SalesReport { DataSource = SampleData.Get() };
    }
}

// ------------------------------------------------------------------
// 3. ViewModel — filtered report with parameter
// ------------------------------------------------------------------
[POCOViewModel]
public class FilteredReportViewModel : ViewModelBase {
    public virtual XtraReport Report { get; protected set; }

    public void ShowForCategory(string category) {
        var report = new SalesReport();
        report.Parameters["Category"].Value = category;
        report.Parameters["Category"].Visible = false;
        Report = report;
    }
}

// ------------------------------------------------------------------
// 4. ReportDesigner Window (WPF license required)
//
// XAML (DesignerWindow.xaml):
//   <Window xmlns:dxrud="http://schemas.devexpress.com/winfx/2008/xaml/reports/userdesigner">
//       <dxrud:ReportDesigner x:Name="reportDesigner" />
//   </Window>
// ------------------------------------------------------------------
public partial class DesignerWindow : Window {
    private void Window_Loaded(object sender, RoutedEventArgs e) {
        // reportDesigner is declared in XAML
        reportDesigner.OpenDocument(new SalesReport());
    }
}

// ------------------------------------------------------------------
// 5. Large report with CachedReportSource
// ------------------------------------------------------------------
public partial class LargeReportWindow : Window {
    private void Window_Loaded(object sender, RoutedEventArgs e) {
        var report = new SalesReport { DataSource = SampleData.GetLarge() };
        var cached = new CachedReportSource(report, new MemoryDocumentStorage());

        // documentPreview declared in XAML as:
        // <dxp:DocumentPreviewControl x:Name="documentPreview" />
        documentPreview.DocumentSource = cached;
        cached.CreateDocumentAsync();   // async — UI stays responsive
    }
}

// ------------------------------------------------------------------
// Sample data / report stubs
// ------------------------------------------------------------------
public class SalesReport : XtraReport {
    public SalesReport() {
        var detail = new DetailBand { HeightF = 20 };
        var label = new XRLabel { BoundsF = new System.Drawing.RectangleF(0, 1, 300, 18) };
        label.ExpressionBindings.Add(new ExpressionBinding("Text", "[Name]"));
        detail.Controls.Add(label);
        Bands.Add(detail);
    }
}

public static class SampleData {
    public record Product(string Name, decimal Price);

    public static System.Collections.Generic.List<Product> Get() => new() {
        new("Widget A", 9.99m),
        new("Widget B", 14.99m),
    };

    public static System.Collections.Generic.List<Product> GetLarge() {
        var list = new System.Collections.Generic.List<Product>();
        for (int i = 0; i < 1000; i++)
            list.Add(new($"Product {i}", i * 1.5m));
        return list;
    }
}

File: plugins/reporting/skills/devexpress-reporting-wpf/examples/quickstart.vb
' DevExpress WPF Reporting — Quickstart (VB.NET)
' Demonstrates: PrintHelper preview, MVVM ViewModel, ReportDesigner

Imports System
Imports System.Windows
Imports DevExpress.Xpf.Printing
Imports DevExpress.Xpf.Reports.UserDesigner
Imports DevExpress.XtraReports.UI
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations

' 1. PrintHelper — simplest approach
Partial Public Class MainWindow
    Inherits Window

    Private Sub btnPreview_Click(sender As Object, e As RoutedEventArgs)
        Dim report As New SalesReport()
        report.DataSource = SampleData.GetProducts()
        PrintHelper.ShowRibbonPrintPreviewDialog(Me, report)
    End Sub
End Class

' 2. MVVM ViewModel for DocumentPreviewControl
'
' XAML:
'   <Window DataContext="{dxmvvm:ViewModelSource Type=local:ReportViewModel}">
'       <dxp:DocumentPreviewControl RequestDocumentCreation="True"
'                                   DocumentSource="{Binding Report}" />
'   </Window>

<POCOViewModel>
Public Class ReportViewModel
    Inherits ViewModelBase

    Public Overridable Property Report As XtraReport

    Protected Overrides Sub OnInitialized()
        LoadReport()
    End Sub

    Public Sub LoadReport()
        Report = New SalesReport() With {.DataSource = SampleData.GetProducts()}
    End Sub
End Class

' 3. ViewModel with parameter
<POCOViewModel>
Public Class FilteredReportViewModel
    Inherits ViewModelBase

    Public Overridable Property Report As XtraReport

    Public Sub ShowForCategory(category As String)
        Dim report As New SalesReport()
        report.Parameters("Category").Value = category
        report.Parameters("Category").Visible = False
        Me.Report = report
    End Sub
End Class

' 4. Designer window (WPF license required)
'
' XAML:
'   <Window xmlns:dxrud="http://schemas.devexpress.com/winfx/2008/xaml/reports/userdesigner">
'       <dxrud:ReportDesigner x:Name="reportDesigner" />
'   </Window>

Partial Public Class DesignerWindow
    Inherits Window

    Private Sub Window_Loaded(sender As Object, e As RoutedEventArgs) Handles Me.Loaded
        reportDesigner.OpenDocument(New SalesReport())
    End Sub
End Class

' Sample report and data
Public Class SalesReport
    Inherits XtraReport

    Public Sub New()
        Dim detail As New DetailBand()
        detail.HeightF = 20

        Dim label As New XRLabel()
        label.BoundsF = New System.Drawing.RectangleF(0, 1, 300, 18)
        label.ExpressionBindings.Add(New ExpressionBinding("Text", "[Name]"))
        detail.Controls.Add(label)
        Bands.Add(detail)
    End Sub
End Class

Public Module SampleData
    Public Class Product
        Public Property Name As String
        Public Property Price As Decimal
    End Class

    Public Function GetProducts() As System.Collections.Generic.List(Of Product)
        Return New System.Collections.Generic.List(Of Product) From {
            New Product() With {.Name = "Widget A", .Price = 9.99D},
            New Product() With {.Name = "Widget B", .Price = 14.99D}
        }
    End Function
End Module
